### PR TITLE
Add hash support to all models that represent parsed objects

### PR DESCRIPTION
**Pull request by ams11**
A followup to https://github.com/workos/workos-ruby/pull/148 - in https://github.com/workos/workos-ruby/pull/148, we added explicit `to_h` methods to `DirectoryUser` and `DirectoryGroup` models - the two models that were affected by the bug reported in https://github.com/workos/workos-ruby/issues/147. Those methods allow converting the classes into a Hash, which customers may be relying on, since the `get_user` and `get_group` end points were previously returning a Hash.

In order to keep the functionality consistent across the rest of our models, adding a module that defines a `to_h` method (which just calls `to_json`) and including this module in each of our models (I went with the ones that also define a `to_json` method, since I figure those are all the ones that behave similarly to `DirectoryUser` and `DirectoryGroup` - did I miss any?). This way, if any customers were relying on treating the results of the `get_user` and `get_group` calls as a Hash, there'll be able to [fairly] easily update their code (by simply inserting a `to_h`)

💓 to @mthadley for helping formulate the plan for both #147 and this PR.

### CODE REVIEW

File: lib/workos/hash_provider.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
+# typed: true
+
+module WorkOS
+  # Module to include an explicit method for converting a model into a Hash containing
+  # its attributes. Default implementation will simply call to_json. Individual classes
+  # may override.
+  module HashProvider
+    include Kernel
```
Comment (mthadley) [inline lib/workos/hash_provider.rb:9]: Do we need to include `Kernel`?

My understanding is that `Kernel` methods are already available everywhere, and so including this module doesn't do anything.

Comment (mthadley) [inline lib/workos/hash_provider.rb:9]: Do we need to include `Kernel`?

My understanding is that `Kernel` methods are already available everywhere, and so including this module doesn't do anything.

Comment (ams11) [reply lib/workos/hash_provider.rb:9]: Ahh, I was trying to remember why I'd included that... I can't remember what it was specifically anymore, but it was for a linter error - it wanted an explicit reference to the definition of some method that it couldn't find here in this module. Probably the `raise` call actually... Let me try and run in through the linter again without `Kernel` and see what it complains about.

Comment (mthadley) [reply lib/workos/hash_provider.rb:9]: Ah, I see. I found a description of the error in the Sorbet docs: https://sorbet.org/docs/error-reference#7003

I guess they recommend doing this, since there's no guarantee that `raise` will be there if the including object inherits directly from `BasicObject`? [The Ruby docs seem to also recommend this when calling Kernel methods in the context of `BasicObject`](https://ruby-doc.org/core-3.0.2/BasicObject.html).

Seems like a corner case, since most classes will inherit from `Object` which includes `Kernel` by default. But, seems fine to follow the advice from Sorbet here.

TLDR: Nevermind, looks fine to me!

**Review by mthadley — APPROVED** [review]

File: lib/workos/hash_provider.rb (hunk 1 above)
Comment (ams11) [reply lib/workos/hash_provider.rb:9]: > Seems like a corner case, since most classes will inherit from Object which includes Kernel by default. But, seems fine to follow the advice from Sorbet here.

Yeah, we can also just disable the Sorbet warning for that line, or call `Kernel.raise` there to be more explicit, but 🤷‍♂️, I don't feel like there's all that much of a difference between any of those options in the end, and [technically, for a very, very edge case of a class that inherits directly from `BasicObject`] including `Kernel` here is safest, so 👍 to just keeping it this way.

> I support another way to implement this is with a base-class, like class Organization < Entity.

Yeah, I considered that also, and honestly don't have a strong preference between the two. Including a module feels just a little bit more flexible, but not by much. I also recall running into some sort of an issue (which I decided not to track down) when I'd tried the base class option, so yeah, as long as you don't feel too strongly about changing this, I'd rather just keeping it as a module to be included.
